Implement time series grouping via HashAggregationOperator using a specialized block hash implementation

martijnvg opened the pull request:
This change adds new operator factory that performs grouping by `_tsid` and `@timestamp` field.

The new operator factory delegates to the `HashAggregationOperator` do the grouping, but does make use of a specialized block hash (which makes a few assumptions about the group fields). This is an initial implementation that is not optimize for 'local' grouping.

The new time series grouping operator factory isn't usable from API and is only usable from tests.

martijnvg committed:
Added WIP time series grouping operator that performs grouping when aggregation mode is final.

martijnvg commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/planner/PlannerUtils.java:
```diff
@@ -66,8 +69,15 @@ public static Tuple<PhysicalPlan, PhysicalPlan> breakPlanBetweenCoordinatorAndDa
-
-            return new ExchangeSourceExec(e.source(), e.output(), e.isInBetweenAggs());
+            List<Attribute> output;
+            if (config.pragmas().timeSeriesMode()) {
```
This was needed because `EsTimeseriesQueryExec` is created after `AbstractPhysicalOperationProviders#groupingPhysicalOperation()` is invoked and `AggregateExec` is created.

I suspect this is the wrong place. The `_tsid` and `@timestamp` fields are needed in both the new grouping operator and the the time series source operator. For the later case `EsTimeseriesQueryExec` was added. 

I wonder whether we can define the need for these fields once in the planner? So that we don't need to indicate in two places (here and in `EsTimeseriesQueryExec`) that these fields are required.

dnhatn replied:
I think we can add these in `AbstractPhysicalOperationProviders#intermediateAttributes`.

martijnvg replied:
I see that this method is used in three places. Only in `groupingPhysicalOperation#pushableStats(...)` there is access to `LocalPhysicalOptimizerContext`, from which query pragma's are accessible. In the other cases this isn't accessible (or we need to push `EsqlConfiguration` / `LocalPhysicalOptimizerContext` down). Or we do we only care about the call to `groupingPhysicalOperation ()` in `groupingPhysicalOperation`?

martijnvg replied:
What do you think of eb00f846c1a550a4a801d4526314c7e04a6ec7cd?
Only in `Mapper#map(...)` I can't figure out whether query pragma's are used. I think I can push it down but requires a lot of plumbing.

dnhatn commented:
Thanks @martijnvg. I wonder if we should consider introducing a new BlockHash implementation that takes advantage of pages sorted by `_tsid`, then `@timestamp`, instead of introducing the TimeSeriesAggregationOperatorFactory.

I also think we need to introduce the rate aggregation function, which is independent of this change. The partial result of the rate function should consist of 5 blocks: earliest timestamp, earliest value, latest timestamp, latest value, and compensation.

martijnvg commented:
>  I wonder if we should consider introducing a new BlockHash implementation that takes advantage of pages sorted by _tsid, then @timestamp, instead of introducing the TimeSeriesAggregationOperatorFactory.

I think a new `BlockHash` implementation makes sense. However I still think we need the `TimeSeriesAggregationOperatorFactory ` here.  The grouping should only take place if the aggregation mode is final, because only then a rate (or other time series statistic) can be computed, once we have all the values for a time serie. Do general intermediate grouping can cause an incorrect rate to be computed. I think the `TimeSeriesAggregationOperatorFactory` can return a `HashAggregationOperator` instance with a 'time series' `BlockHash` implementation.  Does this make sense?

> I also think we need to introduce the rate aggregation function, which is independent of this change. The partial result of the rate function should consist of 5 blocks: earliest timestamp, earliest value, latest timestamp, latest value, and compensation.

👍

martijnvg committed:
Remove custom grouping operator and reuse HashAggregationOperator with
a custom block hash impl.

martijnvg committed:
iter

martijnvg committed:
fixed unit test

martijnvg committed:
Move logic from PlannerUtils#breakPlanBetweenCoordinatorAndDataNode(...) to AbstractPhysicalOperationProviders#.intermediateAttributes(...)

martijnvg committed:
Merge remote-tracking branch 'es/main' into tsdb/compute_engine_ts_grouping_operator

martijnvg committed:
Fix tests after merging in main branch.

martijnvg committed:
Make time series block hash work with new rate function.

Removed the LocalImpl operator, the rate function now keeps all the information around and computes rate when aggregation mode is final.

martijnvg committed:
Moved time series aggregation operator tests from TimeSeriesSortedSourceOperatorTests to TimeSeriesAggregationOperatorTests.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java:
```diff
@@ -0,0 +1,126 @@
+
+    @Override
+    public Block[] getKeys() {
+        int positions = (int) tsidHashes.size();
```
I think the keys should consist of both the tsid and rounded timestamps.

martijnvg replied:
added via: 1ed5b904dabfc299fddde03415fef6de3612b1e0

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java:
```diff
@@ -0,0 +1,126 @@
+            } else {
+                for (int i = 0; i < tsHashVector.getPositionCount(); i++) {
+                    BytesRef tsHash = tsHashVector.getBytesRef(i, spare);
+                    if (tsHash.equals(previousTsidHash) == false) {
```
In the `final` mode, tsid can arrive out of order (from different nodes/shards). We might need to use a `BytesRefLongBlockHash` to hash the grouping ordinals.

martijnvg replied:
done via 1ed5b904dabfc299fddde03415fef6de3612b1e0

martijnvg committed:
added test for grouping with interval and
keep track of group ordinal using BytesRefHash/LongLongHash to avoid incorrect results if aggregation mode is final and results come out of order.

martijnvg committed:
Remove time series interval query pragma and undo planner changes.
This PR will only allow TimeSeriesAggregationOperatorFactory be used in unit tests.

martijnvg committed:
oops, removed too much

elasticsearchmachine commented:
Pinging @elastic/es-storage-engine (Team:StorageEngine)

martijnvg committed:
undo visibility change

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java:
```diff
@@ -0,0 +1,155 @@
+    private final int timestampChannel;
+    private final Rounding.Prepared preparedRounding;
+    private final BytesRefHash tsidHashes;
+    private final LongLongHash intervalHash;
```
++ to use these pair of hashes.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java:
```diff
@@ -0,0 +1,155 @@
+                    BytesRef tsHash = tsHashVector.getBytesRef(i, spare);
+                    long timestampInterval = preparedRounding.round(timestampIntervalBlock.getLong(i));
+                    if (tsHash.equals(previousTsidHash) == false || timestampInterval != previousTimestamp) {
+                        long tsidOrdinal = tsidHashes.add(tsHash);
```
I think we can avoid using these two hashes since data arrives in order in partial mode. Let's keep it simple for now.

martijnvg replied:
Good point, in partial mode we can group like this PR did yesterday (keep track of last seen tsid hash or tsid hash and time stamp and then increment a counter). I can add basically another implementation in a follow-up pr for this later.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java:
```diff
@@ -0,0 +1,155 @@
+
+                        // keep last time stamp around:
+                        long timestamp = timestampIntervalBlock.getLong(i);
+                        intervalHash.add(groupOrdinal, timestamp);
```
I think we miss hashing the rounded timestamps and use it as the `final` groupOrd?

martijnvg replied:
I this case no interval rounding is provided (and `preparedRounding` is null), so I think we can just use the timestamp directly? This is the case when no interval is provided in the es|ql query.

dnhatn replied:
> I this case no interval rounding is provided (and preparedRounding is null), so I think we can just use the timestamp directly? This is the case when no interval is provided in the es|ql query.

@martijnvg I think we might be misunderstanding each other here. In ESQL, there are two modes of grouping aggregation: partial and final (we also have single and intermediate). I think we should specify the rounding for the partial mode, where the raw input includes timestamps and tsids. The output keys for the partial aggregation are the rounded timestamp and tsid.

For the, since the input are the output keys of the partial aggregation, the timestamps are already rounded, so there's no need to specify the rounding. However, we still need to handle the grouping. Using the rounded timestamp directly can lead to a large number of groupings, with many of them being empty.

As for the rounding, we can have an extra eval operator running outside the aggregation instead of inside the block hash, but let's do it in a follow-up.

We can sync up tomorrow to clarify this?

martijnvg replied:
Yes, let's sync about this. I need to think a little bit more about this.

martijnvg replied:
I pushed: 92855af0c78e540c5d269845b975cd11b681d047

The time series grouping code is now much more simplified.
I moved the date rounding to the time series source operator instead of its own dedicated operator.

Let me know if that makes sense. I also did make use of looking up the min/max timestamp, so that a more optimized date rounding can be used.

dnhatn commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java:
```diff
@@ -0,0 +1,155 @@
+        LongVector k2;
+        try (
+            BytesRefVector.Builder tsidHashes = blockFactory.newBytesRefVectorBuilder(positions);
+            LongVector.Builder timestampIntervals = blockFactory.newLongVectorBuilder(positions)
```
nit: we can use `newLongVectorFixedBuilder` instead.

dnhatn reviewed (commented):
Thanks Martijn. I think we are getting closer.

martijnvg committed:
Merge remote-tracking branch 'es/main' into tsdb/compute_engine_ts_grouping_operator

martijnvg committed:
Move ordinal based tests after merging in main.
For now ordinal test don't use TimeSeriesAggregationOperatorFactory.

martijnvg committed:
re-order tests

martijnvg committed:
spotless

martijnvg committed:
newLongVectorFixedBuilder

martijnvg committed:
Merge remote-tracking branch 'es/main' into tsdb/compute_engine_ts_grouping_operator

martijnvg committed:
Move date rounding from TimeSeriesBlockHash to TimeSeriesSortedSourceOperatorFactory.

The added benefit of doing the date rounding in TimeSeriesSortedSourceOperatorFactory,
is that here there is access to timestamp field and we can figure out min and max timestamps.
This allows using (a more optimized) fixed date rounding.

martijnvg committed:
Fix EsTimeseriesQueryExec

dnhatn reviewed (approved):
Can you please update the PR's title? LGTM, thanks Martijn.

martijnvg committed:
iter

martijnvg committed:
Merge remote-tracking branch 'es/main' into tsdb/compute_engine_ts_grouping_operator

martijnvg commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java:
```diff
@@ -0,0 +1,126 @@
+                timestampIntervals.appendLong(intervalHash.getKey2(i));
+            }
+            k1 = tsidHashes.build();
+            k2 = timestampIntervals.build();
```
done: d68f06b492f89109dd0b0afcc2fa3eac8d5112c5

kkrik-es commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/lucene/TimeSeriesSortedSourceOperatorFactory.java:
```diff
@@ -154,15 +189,22 @@ public Page getOutput() {
-                timestampIntervals = timestampIntervalBuilder.build();
-                timestampIntervalBuilder = blockFactory.newLongVectorBuilder(Math.min(remainingDocs, maxPageSize));
-                tsids = tsOrdBuilder.build();
-                tsOrdBuilder = blockFactory.newIntVectorBuilder(Math.min(remainingDocs, maxPageSize));
```
The changes to tsid ordinals should be restored?

kkrik-es replied:
I see, it's included in the tsHashesBuilder.

martijnvg committed:
renamed vars

martijnvg committed:
Merge remote-tracking branch 'es/main' into tsdb/compute_engine_ts_grouping_operator
